lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp

```diff
@@ -0,0 +1,313 @@
+  struct BBState {
+
+    /// Indicates whether a basic block is encountered before seeing a yield
+    /// (BeforeYield) or after seeing a yield (AfterYield), or in both states
+    /// (Conflict). This enum is a semi-lattice where Conflict is the top and
+    /// the merge of BeforeYield and AfterYield states is Conflict.
+    enum YieldState { BeforeYield, AfterYield, Conflict } yieldState;
+
+  private:
+    // The following state is maintained for emitting diagnostics.
+
+    // For AfterYield and Conflict states, this field records the yield
+    // instruction that was seen while propagating the state.
+    SILInstruction *yieldInst;
+
+    BBState(YieldState yState, SILInstruction *yieldI):
+      yieldState(yState), yieldInst(yieldI) { }
+
+  public:
+    static BBState getInitialState() {
+      return BBState(BeforeYield, nullptr);
+    }
+
+    static BBState getAfterYieldState(SILInstruction *yieldI) {
+      assert(yieldI);
+      return BBState(AfterYield, yieldI);
+    }
+
+    static BBState getConflictState(SILInstruction *yieldI) {
+      assert(yieldI);
+      return BBState(Conflict, yieldI);
+    }
+
+    SILInstruction *getYieldInstruction() const {
+      assert(yieldState == AfterYield || yieldState == Conflict);
+      return yieldInst;
+    }
+  };
+
+  /// A structure that records an error found during the analysis along with
+  /// some context information that will be used by diagnostics.
+  struct YieldError {
+    // The kind of error.
+    enum Kind { MultipleYield, ReturnBeforeYield, ReturnOnConflict } errorKind;
+    // The termination instruction where the error should be reported.
+    SILInstruction *termInst;
+    // The input state when the error is encountered.
+    BBState inState;
+
+  private:
+    YieldError(Kind kind, SILInstruction *term, BBState state)
+        : errorKind(kind), termInst(term), inState(state) {}
+
+  public:
+    static YieldError getMultipleYieldError(YieldInst *yield, BBState state) {
+      assert(state.yieldState != BBState::BeforeYield);
+      return YieldError(MultipleYield, yield, state);
+    }
+
+    static YieldError getReturnBeforeYieldError(ReturnInst *returnI,
+                                                BBState state) {
+      assert(state.yieldState == BBState::BeforeYield);
+      return YieldError(ReturnBeforeYield, returnI, state);
+    }
+
+    static YieldError getReturnOnConflict(ReturnInst *returnI, BBState state) {
+      assert(state.yieldState == BBState::Conflict);
+      return YieldError(ReturnOnConflict, returnI, state);
+    }
+  };
+
+  /// Transfer function of the data-flow analysis.
+  ///
+  /// \param bb Basic block that should be processed
+  /// \param inState BBState at the start of the basic block
+  /// \param error out parameter that will contain information about
+  /// an error that is detected.
+  /// \return the state at the exit of the basic block if it can be computed
+  /// and None otherwise.
+  llvm::Optional<BBState>
```

Comment:
We have imported Optional into the swift namespace, so I don't think you need the llvm:: any more for it. On another note, can you also make this helper function have static on it?